betty/concurrent.py

```diff
@@ -2,49 +2,112 @@
 Provide utilities for concurrent programming.
 """
 import asyncio
+import threading
 import time
 from asyncio import sleep
 from math import floor
 from threading import Lock
 from types import TracebackType
+from typing import Self
 
+from betty.asyncio import gather
 
-class AsynchronizedLock:
+
+class _Lock:
     """
-    Make a sychronous (blocking) lock asynchronous (non-blocking).
+    Provide an asynchronous lock.
     """
 
-    def __init__(self, lock: Lock):
-        self._lock = lock
-
     async def __aenter__(self):
         await self.acquire()
 
     async def __aexit__(self, exc_type: type[BaseException] | None, exc_val: BaseException | None, exc_tb: TracebackType | None) -> None:
         self.release()
 
-    async def acquire(self) -> None:
-        while not self._lock.acquire(False):
-            # Sleeping for zero seconds does not actually sleep, but gives the event
-            # loop a chance to progress other tasks while we wait for another chance
-            # to acquire the lock.
-            await sleep(0)
+    async def acquire(self, *, wait: bool = True) -> bool:
+        raise NotImplementedError
+
+    def release(self) -> None:
+        raise NotImplementedError
+
+
+async def asynchronize_acquire(lock: Lock, *, wait: bool = True) -> bool:
+    """
+    Acquire a synchronous lock asynchronously.
+    """
+    while not lock.acquire(blocking=False):
+        if not wait:
+            return False
+        # Sleeping for zero seconds does not actually sleep, but gives the event
+        # loop a chance to progress other tasks while we wait for another chance
+        # to acquire the lock.
+        await sleep(0)
+    return True
+
+
+class AsynchronizedLock(_Lock):
+    """
+    Make a sychronous (blocking) lock asynchronous (non-blocking).
+    """
+
+    __slots__ = '_lock'
+
+    def __init__(self, lock: Lock):
+        self._lock = lock
+
+    async def acquire(self, *, wait: bool = True) -> bool:
+        return await asynchronize_acquire(self._lock, wait=wait)
 
     def release(self) -> None:
         self._lock.release()
 
+    @classmethod
+    def threading(cls) -> Self:
+        return cls(threading.Lock())
+
+
+class MultiLock(_Lock):
+    """
+    Provide a lock that only acquires if all of the given locks can be acquired.
+    """
+
+    __slots__ = '_locked', '_locks'
+
+    def __init__(self, *locks: _Lock):
+        self._locks = locks
+        self._locked = False
+
+    async def acquire(self, *, wait: bool = True) -> bool:
+        acquisitions = await gather(*(lock.acquire(wait=wait) for lock in self._locks))
```

Comment:
If this returns tuples, we won't have to `zip()` afterwards.

Reply:
But then we wouldn't easily be able to check if any acquisition failed. Never mind!